lib/grammar.js

```diff
@@ -0,0 +1,328 @@
+'use strict';
+
+/**
+ * Removes `[`, `]` and any trailing quantifiers from the source of a RegExp
+ * @param {RegExp} regexp
+ */
+function chars(regexp) {
+	if (regexp.source[0] !== '[') {
+		throw new Error(regexp + ' can not be used with chars');
+	}
+	return regexp.source.slice(1, regexp.source.lastIndexOf(']'));
+}
+
+/**
+ * Creates a new character list regular expression,
+ * by removing `search` from the source of `regexp`.
+ * @param {RegExp} regexp
+ * @param {string} search the characters to remove
+ * @returns {RegExp & {chars:string}}
+ */
+function chars_without(regexp, search) {
+	if (regexp.source[0] !== '[') {
+		throw new Error('/' + regexp.source + '/ can not be used with chars_without');
+	}
+	if (!search) {
+		throw new Error(search + ' is falsy');
+	}
+	if (regexp.source.indexOf(search) === -1) {
+		throw new Error('"' + search + '" is not is /' + regexp.source + '/');
+	}
+	var result = new RegExp(regexp.source.replace(search, ''));
+	result.chars = chars(result);
+	return result;
+}
+
+/**
+ * Combines and Regular expressions correctly by using `RegExp.source`.
+ * @param {...(RegExp | string)[]} args
+ * @returns {RegExp}
+ */
+function reg(args) {
+	var self = this;
+	return new RegExp(
+		Array.prototype.slice
+			.call(arguments)
+			.map(function (part, index) {
+				var isStr = typeof part === 'string';
+				if (isStr && self === undefined && part === '|') {
+					throw new Error('use regg instead of reg to wrap expressions with `|`!');
+				}
+				return isStr ? part : part.source;
+			})
+			.join(''),
+		'm'
+	);
+}
+
+/**
+ * Like `reg` but adds a `chars` property by applying the function with the same name.
+ * @param {...(RegExp | string)[]} args
+ * @returns {RegExp & {chars:string}}
+ */
+function reg_chars(args) {
+	// var args = Array.prototype.slice.call(arguments);
+	var result = reg.apply(reg_chars, arguments);
+	result.chars = chars(result);
+	return result;
+}
+
+/**
+ * Like `reg` but wraps the expression in `(?:`,`)` to create a non tracking group.
+ * @param {...(RegExp | string)[]} args
+ * @returns {RegExp}
+ */
+function regg(args) {
+	if (arguments.length === 0) {
+		throw new Error('no parameters provided');
+	}
+	return reg.apply(regg, ['(?:'].concat(Array.prototype.slice.call(arguments), [')']));
+}
+
+// https://www.w3.org/TR/xml/#document
+// [1] document ::= prolog element Misc*
+// https://www.w3.org/TR/xml11/#NT-document
+// [1] document    ::= ( prolog element Misc* ) - ( Char* RestrictedChar Char* )
+
+// https://www.w3.org/TR/xml/#char32
+// https://www.w3.org/TR/xml11/#NT-Char
+// any Unicode character, excluding the surrogate blocks, FFFE, and FFFF.
+// [2] Char ::= #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]
+var Char = /[\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD]/; // | [#x10000-#x10FFFF]
```

Comment:
```suggestion
var Char = /[-\u0001-\u002D\u002E-\uD7FF\uE000-\uFFFD]/; // | [#x10000-#x10FFFF]
```

Fix for `chars_without(Char,'-');`
We should find a way to check for the surrogate pair as well